### PR TITLE
[ScatterOp 4] ScatterOp IndexLowering (different shape case)

### PR DESCRIPTION
**Pull request by ftxj**
This PR support the codegen for `ScatterOp` when the `inputTv` and `indexTv` has different size. 

The `index lowering` and `predicate generate` passes have different behavior between `ScatterOp` and other ops. 

For example, other ops generate the `loop bounds`, `predicate start` by using the extends of `output_tv`. But for `ScatterOp`, we need to use the extent of `index_tv`. To do this, we modify the `compute_at_map`, to make the concrete_id of `output_tv` iteration domain is `index_tv` iteration domain.

Due to the `IndexCompute` is very complicated, the implementation of the above logic is very hack. It's not a good way to do this. I think maybe we need a more general method to deal with different shape between different tensor view. I look forward to the view of reviewers in this point.

Another thing is that the `ScatterOp` need to initialize the `output_tv` by using the `input_tv`, so we add a method `getTensorForFillAnotherTensor` to do this.

### CODE REVIEW

**Comment by naoyam** [comment]
Thanks for the PR. Looks like the implication of the changes seems to be rather significant and doesn't seem to align well with the current system. This is not your fault at all, but it clearly exposes one of the major limitations of nvfuser. While I'm glad to see you were able to work around the limitation even for a limited case, I strongly feel we really need to build something more fundamental.

Let me try to see if the change could be minimized.

File: third_party/nvfuser/csrc/compute_at_map.cpp [hunk 1]
```diff
@@ -1106,6 +1127,9 @@ void ComputeAtMap::buildConcreteIds() {
         "Cannot compute concrete id of empty set.");
     auto first_id = disjoint_set_shared_ptr->vector().front();
     auto concrete_id = computeConcreteId(first_id, IdMappingMode::PERMISSIVE);
+    if (concrete_id_map_.find(concrete_id) != concrete_id_map_.end()) {
```
Comment (naoyam) [inline third_party/nvfuser/csrc/compute_at_map.cpp]: `computeConcreteId` is also modifed to use `concrete_id_map_`. Is the change here also necessary? Same for the `ALMOSTEXACT` and `LOOP` maps.

I tried commenting out them and didn't see any error with `FusionScatter2DZerosSelfTvFusion_CUDA`.

File: third_party/nvfuser/csrc/ir_utils.h [hunk 2]
```diff
@@ -310,6 +310,8 @@ TORCH_CUDA_CU_API std::vector<IndexSelectOp*> getIndexSelectOps(Fusion* fusion);
 
 TORCH_CUDA_CU_API std::vector<TorchGatherOp*> getTorchGatherOps(Fusion* fusion);
 
+TORCH_CUDA_CU_API std::vector<ScatterOp*> getScatterOps(Fusion* fusion);
```
Comment (naoyam) [inline third_party/nvfuser/csrc/ir_utils.h:313]: nit: I'm not strongly against, but we shouldn't create a `getXYZOps` for every type of ops as the code is fairly trivial.

File: third_party/nvfuser/csrc/compute_at_map.cpp [hunk 3]
```diff
@@ -1088,12 +1092,29 @@ void ComputeAtMap::buildConcreteIds() {
...
+      concrete_id_map_[src_id] = idx_id;
+    }
+  }
   for (const auto& disjoint_set_shared_ptr :
        id_graph_.exactNodes().disjointSets()) {
     TORCH_INTERNAL_ASSERT(
         disjoint_set_shared_ptr->vector().size(),
         "Cannot compute concrete id of empty set.");
     auto first_id = disjoint_set_shared_ptr->vector().front();
+    if (concrete_id_map_.find(first_id) != concrete_id_map_.end()) {
```
Comment (naoyam) [inline third_party/nvfuser/csrc/compute_at_map.cpp]: Is this really necessary? I don't know why the concrete ID of the exact map needs to be changed.

Commented out and ran the C++ test. Seems to working fine.

**Comment by naoyam** [comment]
Actually, while I don't see any test failure, compute-sanitizer shows write errors. Run the test as:

```
PYTORCH_NVFUSER_DUMP=debug_info PYTORCH_NO_CUDA_MEMORY_CACHING=1 compute-sanitizer ./build/bin/nvfuser_tests --gtest_filter="*FusionScatter2DZerosSelfTvFusion_CUDA" 
```

```
PRINTING: __tmp_kernel1.cu
check cg_outputs ========= Invalid __global__ write of size 4 bytes
=========     at 0x290 in /raid/nmaruyama/debug1/__tmp_kernel1.cu:9192:CudaCodeGen::kernel1(CudaCodeGen::Tensor<float, (int)2>, CudaCodeGen::Tensor<long long, (int)2>, CudaCodeGen::Tensor<long long, (int)2>, CudaCodeGen::Tensor<float, (
int)2>, CudaCodeGen::Tensor<float, (int)2>)
=========     by thread (96,0,0) in block (0,0,0)
=========     Address 0x7fdbda600f80 is out of bounds
=========     and is 337 bytes after the nearest allocation at 0x7fdbda600e00 of size 48 bytes
=========     Saved host backtrace up to driver entry point at kernel launch time
=========     Host Frame: [0x304e32]
=========                in /usr/lib/x86_64-linux-gnu/libcuda.so.1
```

Note the error happens with no modification on A100. It seems there's something wrong, probably predicates?

File: third_party/nvfuser/csrc/compute_at_map.cpp (hunk 3 above)
Comment (ftxj) [reply third_party/nvfuser/csrc/compute_at_map.cpp]: If we don't change concrete ID of exact map, the index of `output_tv` will be wrong. We want to generate index like this `T_output[threadIdx.x % T_index.size(0) + ...]]`. Notice that we use stride `T_index.size(0)` to access `T_output`. I have add more test cases in new commit.

File: third_party/nvfuser/csrc/compute_at_map.cpp (hunk 1 above)
Comment (ftxj) [reply third_party/nvfuser/csrc/compute_at_map.cpp]: If we don't change `ALMOSTEXACT`. The vectorize pass will give some failure.  For `LOOP` map, we need to generate loop according to the `tv_index`, not `tv_output` or `tv_input`. So, we need to change it. I have added test case `FusionScatterNoParallism_CUDA` and delete some unused logic to show this.

**Comment by ftxj** [comment]
> PYTORCH_NO_CUDA_MEMORY_CACHING=1

I will try to fix it.

**Comment by ftxj** [comment]
> Actually, while I don't see any test failure, compute-sanitizer shows write errors. Run the test as:
> 
> ```
> PYTORCH_NVFUSER_DUMP=debug_info PYTORCH_NO_CUDA_MEMORY_CACHING=1 compute-sanitizer ./build/bin/nvfuser_tests --gtest_filter="*FusionScatter2DZerosSelfTvFusion_CUDA" 
> ```
> 
> ```
> PRINTING: __tmp_kernel1.cu
> check cg_outputs ========= Invalid __global__ write of size 4 bytes
> =========     at 0x290 in /raid/nmaruyama/debug1/__tmp_kernel1.cu:9192:CudaCodeGen::kernel1(CudaCodeGen::Tensor<float, (int)2>, CudaCodeGen::Tensor<long long, (int)2>, CudaCodeGen::Tensor<long long, (int)2>, CudaCodeGen::Tensor<float, (
> int)2>, CudaCodeGen::Tensor<float, (int)2>)
> =========     by thread (96,0,0) in block (0,0,0)
> =========     Address 0x7fdbda600f80 is out of bounds
> =========     and is 337 bytes after the nearest allocation at 0x7fdbda600e00 of size 48 bytes
> =========     Saved host backtrace up to driver entry point at kernel launch time
> =========     Host Frame: [0x304e32]
> =========                in /usr/lib/x86_64-linux-gnu/libcuda.so.1
> ```
> 
> Note the error happens with no modification on A100. It seems there's something wrong, probably predicates?

I will try to fix it.

**Comment by ftxj** [comment]
> Actually, while I don't see any test failure, compute-sanitizer shows write errors. Run the test as:
> 
> ```
> PYTORCH_NVFUSER_DUMP=debug_info PYTORCH_NO_CUDA_MEMORY_CACHING=1 compute-sanitizer ./build/bin/nvfuser_tests --gtest_filter="*FusionScatter2DZerosSelfTvFusion_CUDA" 
> ```
> 
> ```
> PRINTING: __tmp_kernel1.cu
> check cg_outputs ========= Invalid __global__ write of size 4 bytes
> =========     at 0x290 in /raid/nmaruyama/debug1/__tmp_kernel1.cu:9192:CudaCodeGen::kernel1(CudaCodeGen::Tensor<float, (int)2>, CudaCodeGen::Tensor<long long, (int)2>, CudaCodeGen::Tensor<long long, (int)2>, CudaCodeGen::Tensor<float, (
> int)2>, CudaCodeGen::Tensor<float, (int)2>)
> =========     by thread (96,0,0) in block (0,0,0)
> =========     Address 0x7fdbda600f80 is out of bounds
> =========     and is 337 bytes after the nearest allocation at 0x7fdbda600e00 of size 48 bytes
> =========     Saved host backtrace up to driver entry point at kernel launch time
> =========     Host Frame: [0x304e32]
> =========                in /usr/lib/x86_64-linux-gnu/libcuda.so.1
> ```
> 
> Note the error happens with no modification on A100. It seems there's something wrong, probably predicates?

After a long time, I have fixed this bug..

**Comment by naoyam** [comment]
Since we haven't started reviewing this PR fully yet, can you move this to the new repository? github.com/NVIDIA/Fuser

**Comment by naoyam** [comment]
Closing this one in favor of https://github.com/NVIDIA/Fuser/pull/89

**Pull request closed without merging**